MSA-757: Timevalve Smart

SmartThings opened the pull request:
Submitting 'Timevalve Smart' for publication

Commit:
MSA-757: Smart Gas-Lock work with z-wave

dkirker commented on devicetypes/timevalve-gaslock-t-08/timevalve-smart.src/timevalve-smart.groovy:
```diff
@@ -0,0 +1,281 @@
+    return result
+}
+
+// remote open not allow
```
If open is not allowed to be supported for remote control, then the open() function should be empty. Otherwise, SmartApps will be able to call open() on the device and execute open().

ruinnel replied:
remove all code in open()

jasonbio commented on devicetypes/timevalve-gaslock-t-08/timevalve-smart.src/timevalve-smart.groovy:
```diff
@@ -0,0 +1,281 @@
+// If you add the Configuration capability to your device type, this
+// command will be called right after the device joins to set
+// device-specific configuration commands.
+def configure() {
```
empty function definition with Configuration capability present

ruinnel replied:
remove capability and command.

jasonbio commented on devicetypes/timevalve-gaslock-t-08/timevalve-smart.src/timevalve-smart.groovy:
```diff
@@ -0,0 +1,281 @@
+
+            main (["statusTile"])
+//            details (["statusTile", "remainingSliderTile", "setRemaining", "setTimeout10", "setTimeout20", "batteryTile", "refreshTile", "setTimeout30", "setTimeout40"])
+//            details (["statusTile", "batteryTile", "setRemaining", "refreshTile"])
```
Was setRemaining ever meant to be called? Just double checking, since it's defined just not tied to any action

ruinnel replied:
yes. setRemaining never called.
several function and tile hide by intention.
function is operating well.
However, it not reached the UI requirements.

jasonbio commented on devicetypes/timevalve-gaslock-t-08/timevalve-smart.src/timevalve-smart.groovy:
```diff
@@ -0,0 +1,281 @@
+                attribute "remaining", "number"
+                attribute "remainingText", "String"
+                attribute "timeout", "number"
+        }
```
No fingerprint definition for this device?

tylerlange replied:
To clarify - with no fingerprint, this Device Type Handler will never be assigned to the valve when it's paired.

ruinnel replied:
add fingerprint
fingerprint deviceId:"0x1006", inClusters:"0x5E, 0x86, 0x72, 0x5A, 0x73, 0x98, 0x80"

tylerlange replied:
@ruinnel thanks for including the fingerprint. You'll need to also update the submission/pull request.

Have you tested this code with the actual device (including fingerprint)?

jasonbio commented on devicetypes/timevalve-gaslock-t-08/timevalve-smart.src/timevalve-smart.groovy:
```diff
@@ -0,0 +1,281 @@
+metadata {
+        definition (name: "Timevalve Smart", namespace: "timevalve.gaslock.t-08", author: "ruinnel") {
+                capability "Valve"
```
Format the name differently, ex: "Smart TimeValve Gaslock"

kris-schaller replied:
Note: If the name of the smartapp changes then the file name needs to change as well, which last I heard doesn't update. This may require a new SmartApp creation.

ruinnel replied:
"TimeValve Smart" was product name.

dkirker replied:
If that is the product name, that is fine.

jasonbio replied:
Format with correct branding then (capital "V") "TimeValve Smart"

jasonbio commented on devicetypes/timevalve-gaslock-t-08/timevalve-smart.src/timevalve-smart.groovy:
```diff
@@ -0,0 +1,281 @@
+        }
+
+        simulator {
+        }
```
empty simulator declaration

ruinnel replied:
remove - empty simulator declearation

jasonbio commented on devicetypes/timevalve-gaslock-t-08/timevalve-smart.src/timevalve-smart.groovy:
```diff
@@ -0,0 +1,281 @@
+    	result.add( createEvent(name: "timeout", value: value, displayed: false, isStateChange: true) )
+    }
+    
+    /*
```
consider removing if this commented block isn't needed

ruinnel replied:
commented block removed...

dkirker commented on devicetypes/timevalve-gaslock-t-08/timevalve-smart.src/timevalve-smart.groovy:
```diff
@@ -0,0 +1,281 @@
+	def type = cmd.notificationType
+    if (type == cmd.NOTIFICATION_TYPE_HEAT) {
+    	log.info "NotificationReport - ${type}"
+        createEvent(name: "temperature", value: 80, descriptionText: "${device.displayName} is over heat!", displayed: true, isStateChange: true)
```
It might be more appropriate to set a larger value (like 999), and then define the Units for this. Such as:

createEvent(name: "temperature", value: 999, unit: "C", descriptionText: "${device.displayName} is over heat!", displayed: true, isStateChange: true)

ruinnel replied:
I change to

createEvent(name: "temperature", value: 999, unit: "C", descriptionText: msg, displayed: true, isStateChange: true)

jasonbio commented on devicetypes/timevalve-gaslock-t-08/timevalve-smart.src/timevalve-smart.groovy:
```diff
@@ -0,0 +1,281 @@
+                capability "Notification"
+				capability "Configuration"
+                capability "Temperature Measurement"
+                
```
Consider adding logic for auto temperature localization based on hub settings (Celsius vs Fahrenheit) and proper conversion for accurate overheat detection

ruinnel replied:
"Temperature Measurement" are using for the notification function. Temperature value is not important.
and.. I can't find proper capability.

jasonbio replied:
That's fine, thanks for clarification

dkirker commented on devicetypes/timevalve-gaslock-t-08/timevalve-smart.src/timevalve-smart.groovy:
```diff
@@ -0,0 +1,281 @@
+
+        tiles (scale: 2) {
+        	multiAttributeTile(name:"statusTile", type:"thermostat", width:6, height:4) {
+            	tileAttribute("device.contact", key: "PRIMARY_CONTROL") {
```
Should be device.valve

ruinnel replied:
device.valve - not work.

dkirker replied:
If you make the changes that I have indicated relating to this and the multiAttributeTile it will.

dkirker commented on devicetypes/timevalve-gaslock-t-08/timevalve-smart.src/timevalve-smart.groovy:
```diff
@@ -0,0 +1,281 @@
+
+def zwaveEvent(physicalgraph.zwave.commands.switchbinaryv1.SwitchBinaryReport cmd) {
+	//log.debug "switch status - " + cmd.value
+    createEvent(name:"contact", value: cmd.value ? "open" : "closed")
```
Should be valve and not contact

ruinnel replied:
valve is capability name.
createEvent 'name' need attribute name. right?
now work fine.

dkirker commented on devicetypes/timevalve-gaslock-t-08/timevalve-smart.src/timevalve-smart.groovy:
```diff
@@ -0,0 +1,281 @@
+    	value = 0
+    }
+    
+    if (device.currentValue('contact') == 'open') {
```
Change contact to valve.

ruinnel replied:
nope.
device.currentValue need 'attribute' name.
valve is capability name.
'contact' is valve's attribute.

dkirker replied:
It seems that our documentation and device examples are not consistent, so I am asking our platform engineers to clarify.

dkirker replied:
I verified and it is contact. So as long as all of the attribute names reference contact it will work.

dkirker commented on devicetypes/timevalve-gaslock-t-08/timevalve-smart.src/timevalve-smart.groovy:
```diff
@@ -0,0 +1,281 @@
+        }
+
+        tiles (scale: 2) {
+        	multiAttributeTile(name:"statusTile", type:"thermostat", width:6, height:4) {
```
thermostat is incorrect. Should be generic

dkirker commented on devicetypes/timevalve-gaslock-t-08/timevalve-smart.src/timevalve-smart.groovy:
```diff
@@ -0,0 +1,281 @@
+	log.info "zwave.configurationV1.configurationGet - " + cmd
+    def array = cmd.configurationValue
+    def value = ( (array[0] * 0x1000000) + (array[1] * 0x10000) + (array[2] * 0x100) + array[3] ).intdiv(60)
+    if (device.currentValue("switch") == "on") {
```
Should this be valve == "open"?

ruinnel replied:
change to 
 if (device.currentValue("contact") == "open") {

dkirker commented on devicetypes/timevalve-gaslock-t-08/timevalve-smart.src/timevalve-smart.groovy:
```diff
@@ -0,0 +1,281 @@
+    	value = ( (array[0] * 0x1000000) + (array[1] * 0x10000) + (array[2] * 0x100) + array[3] ).intdiv(60)
+    } else {
+    	value = 0
+    }
```
Cleaner code would be:

```
def value = 0

if (device.currentValue("valve") == "open") {
    value = ( (array[0] * 0x1000000) + (array[1] * 0x10000) + (array[2] * 0x100) + array[3] ).intdiv(60)
}
```

dkirker replied:
Actually, this can all just be consolidated to 

```
def value = 0
if (device.currentValue("contact") == "open") {
    value = ( (array[0] * 0x1000000) + (array[1] * 0x10000) + (array[2] * 0x100) + array[3] ).intdiv(60)

    def hour = value.intdiv(60);
    def min = (value % 60).toString().padLeft(2, '0');
    def text = "${hour}:${min}M"
```

...

jasonbio commented:
Checking in to see when we can expect a new commit with updated code - stuck in code review until then. Thanks!

Commit:
Modifying 'Timevalve Smart'

jasonbio committed:
Update timevalve-smart.groovy

Commenting out fingerprint temporarily to avoid potential conflicts with other devices as this devices is specifically for a Korean deployment in AP01 - see DVCSMP-1425

jasonbio commented:
This PR is being merged specifically for deployment in AP01 due to partner request, fingerprint has been commented out to avoid potential conflict (DVCSMP-1425)
